File: src/main/java/net/guizhanss/fastmachines/items/machines/generic/AbstractFastMachine.java
package net.guizhanss.fastmachines.items.machines.generic;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.annotation.Nonnull;
import javax.annotation.ParametersAreNonnullByDefault;

import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.inventory.ItemStack;

import io.github.thebusybiscuit.slimefun4.api.SlimefunAddon;
import io.github.thebusybiscuit.slimefun4.api.items.SlimefunItemStack;
import io.github.thebusybiscuit.slimefun4.api.items.settings.IntRangeSetting;
import io.github.thebusybiscuit.slimefun4.api.recipes.RecipeType;
import io.github.thebusybiscuit.slimefun4.core.attributes.EnergyNetComponent;
import io.github.thebusybiscuit.slimefun4.core.networks.energy.EnergyNetComponentType;
import io.github.thebusybiscuit.slimefun4.libraries.dough.blocks.BlockPosition;
import io.github.thebusybiscuit.slimefun4.libraries.dough.items.CustomItemStack;
import io.github.thebusybiscuit.slimefun4.utils.ChestMenuUtils;
import io.github.thebusybiscuit.slimefun4.utils.HeadTexture;
import io.github.thebusybiscuit.slimefun4.utils.LoreBuilder;

import me.mrCookieSlime.Slimefun.api.inventory.BlockMenu;
import me.mrCookieSlime.Slimefun.api.inventory.BlockMenuPreset;

import net.guizhanss.fastmachines.FastMachines;
import net.guizhanss.fastmachines.core.items.attributes.NotAHopper;
import net.guizhanss.fastmachines.core.recipes.IRecipe;
import net.guizhanss.fastmachines.setup.Groups;
import net.guizhanss.fastmachines.utils.Heads;
import net.guizhanss.guizhanlib.slimefun.machines.TickingMenuBlock;

import lombok.Getter;

/**
 * A fast machine is a machine that craft items without putting ingredients in shape, and can bulk craft.
 *
 * @author ybw0014
 */
@SuppressWarnings("ConstantConditions")
public abstract class AbstractFastMachine extends TickingMenuBlock implements EnergyNetComponent, NotAHopper {

    // slots
    // @formatter:off
    static final int[] INPUT_SLOTS = new int[] {
        0, 1, 2, 3, 4, 5, 6, 7, 8,
        9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20, 21, 22, 23, 24, 25, 26,
        27, 28, 29, 30, 31, 32, 33, 34, 35
    };
    static final int[] OUTPUT_SLOTS = new int[] {
        27, 28, 29, 30, 31, 32, 33, 34, 35,
        18, 19, 20, 21, 22, 23, 24, 25, 26,
        9, 10, 11, 12, 13, 14, 15, 16, 17,
        0, 1, 2, 3, 4, 5, 6, 7, 8,
    };
    static final int[] PREVIEW_SLOTS = new int[] {
        36, 37, 38, 39, 40, 41,
        45, 46, 47, 48, 49, 50
    };
    // @formatter:on
    static final int SCROLL_UP_SLOT = 42;
    static final int SCROLL_DOWN_SLOT = 51;
    static final int CHOICE_SLOT = 52;
    static final int CRAFT_SLOT = 53;
    static final int INFO_SLOT = 43;
    static final int ENERGY_SLOT = 44;
    // constants
    static final int ITEMS_PER_PAGE = PREVIEW_SLOTS.length;
    // menu items
    static final ItemStack NO_ITEM = FastMachines.getLocalization().getItem("NO_ITEM", Material.BARRIER);
    static final ItemStack SCROLL_UP_ITEM = FastMachines.getLocalization().getItem("SCROLL_UP", Heads.ARROW_UP.getTexture());
    static final ItemStack SCROLL_DOWN_ITEM = FastMachines.getLocalization().getItem("SCROLL_DOWN", Heads.ARROW_DOWN.getTexture());
    static final ItemStack INFO_ITEM = FastMachines.getLocalization().getItem("INFO", Heads.INFO.getTexture());

    @Getter
    protected final List<IRecipe> recipes = new ArrayList<>();

    private final IntRangeSetting energyPerUseSetting;
    private final IntRangeSetting energyCapacitySetting;
    private final Map<BlockPosition, FastMachineCache> caches = new HashMap<>();

    @ParametersAreNonnullByDefault
    protected AbstractFastMachine(SlimefunItemStack item, RecipeType recipeType, ItemStack[] recipe, int energyPerUse, int energyCapacity) {
        super(Groups.MACHINES, item, recipeType, recipe);

        energyPerUseSetting = new IntRangeSetting(this, "energy-per-use", 0, energyPerUse, Integer.MAX_VALUE);
        energyCapacitySetting = new IntRangeSetting(this, "energy-capacity", 0, energyCapacity, Integer.MAX_VALUE);

        addItemSetting(energyPerUseSetting, energyCapacitySetting);
    }

    @Nonnull
    @Override
    public EnergyNetComponentType getEnergyComponentType() {
        return EnergyNetComponentType.CONSUMER;
    }

    public int getEnergyPerUse() {
        return FastMachines.getConfigService().isFastMachinesUseEnergy() ? energyPerUseSetting.getValue() : 0;
    }

    @Override
    public int getCapacity() {
        return FastMachines.getConfigService().isFastMachinesUseEnergy() ? energyCapacitySetting.getValue() : 0;
    }

    @Override
    protected void setup(@Nonnull BlockMenuPreset preset) {
        for (int slot : PREVIEW_SLOTS) {
            preset.addItem(slot, ChestMenuUtils.getBackground(), ChestMenuUtils.getEmptyClickHandler());
        }
        preset.addItem(INFO_SLOT, INFO_ITEM, ChestMenuUtils.getEmptyClickHandler());
        preset.addItem(CHOICE_SLOT, NO_ITEM, ChestMenuUtils.getEmptyClickHandler());
        preset.addItem(SCROLL_UP_SLOT, SCROLL_UP_ITEM, ChestMenuUtils.getEmptyClickHandler());
        preset.addItem(SCROLL_DOWN_SLOT, SCROLL_DOWN_ITEM, ChestMenuUtils.getEmptyClickHandler());

        preset.addItem(CRAFT_SLOT, getCraftItem(), ChestMenuUtils.getEmptyClickHandler());
        preset.addItem(ENERGY_SLOT, new CustomItemStack(HeadTexture.ENERGY_CONNECTOR.getAsItemStack(), " ", LoreBuilder.power(getEnergyPerUse(), FastMachines.getLocalization().getString("lores.per-craft"))), ChestMenuUtils.getEmptyClickHandler());
    }

    @Override
    @ParametersAreNonnullByDefault
    protected void onNewInstance(BlockMenu menu, Block block) {
        BlockPosition pos = new BlockPosition(block);
        caches.put(pos, new FastMachineCache(this, menu));
    }

    @Override
    @ParametersAreNonnullByDefault
    protected void tick(Block b, BlockMenu menu) {
        BlockPosition pos = new BlockPosition(b);
        if (menu.hasViewer() && caches.containsKey(pos)) {
            caches.get(pos).tick();
        }
    }

    @Override
    public int[] getInputSlots() {
        return INPUT_SLOTS;
    }

    @Override
    public int[] getOutputSlots() {
        return new int[0];
    }

    @Override
    public void register(@Nonnull SlimefunAddon addon) {
        super.register(addon);
        if (!this.isDisabled()) {
            FastMachines.getRegistry().getEnabledFastMachines().add(this);
        }
    }

    @Override
    @ParametersAreNonnullByDefault
    protected void onBreak(BlockBreakEvent e, BlockMenu menu) {
        super.onBreak(e, menu);
        Location l = menu.getLocation();
        menu.dropItems(l, INPUT_SLOTS);
        caches.remove(new BlockPosition(l));
    }

    /**
     * Register available recipes for this machine.
     * <p>
     * Note: this method is called synchronously after Slimefun item registry finishes loading.
     */
    public abstract void registerRecipes();

    /**
     * Get the item of the craft button of this machine.
     *
     * @return the craft button item.
     */
    protected abstract ItemStack getCraftItem();
}

File: src/main/kotlin/net/guizhanss/fastmachines/implementation/items/machines/generic/AbstractFastMachine.kt
package net.guizhanss.fastmachines.implementation.items.machines.generic

import io.github.thebusybiscuit.slimefun4.api.items.ItemGroup
import io.github.thebusybiscuit.slimefun4.api.items.SlimefunItemStack
import io.github.thebusybiscuit.slimefun4.api.items.settings.IntRangeSetting
import io.github.thebusybiscuit.slimefun4.api.recipes.RecipeType
import io.github.thebusybiscuit.slimefun4.core.attributes.EnergyNetComponent
import io.github.thebusybiscuit.slimefun4.core.networks.energy.EnergyNetComponentType
import io.github.thebusybiscuit.slimefun4.libraries.dough.blocks.BlockPosition
import io.github.thebusybiscuit.slimefun4.utils.ChestMenuUtils
import io.github.thebusybiscuit.slimefun4.utils.LoreBuilder
import me.mrCookieSlime.Slimefun.api.inventory.BlockMenu
import me.mrCookieSlime.Slimefun.api.inventory.BlockMenuPreset
import net.guizhanss.fastmachines.FastMachines
import net.guizhanss.fastmachines.core.FMRegistry
import net.guizhanss.fastmachines.core.recipes.Recipe
import net.guizhanss.fastmachines.core.recipes.loaders.RecipeLoader
import net.guizhanss.fastmachines.utils.constants.HeadTexture
import net.guizhanss.fastmachines.utils.items.withLore
import net.guizhanss.fastmachines.utils.items.withName
import net.guizhanss.guizhanlib.slimefun.machines.MenuBlock
import org.bukkit.Material
import org.bukkit.block.Block
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.inventory.ItemStack
import io.github.thebusybiscuit.slimefun4.utils.HeadTexture as SlimefunHeadTexture

/**
 * The base fast machine.
 */
abstract class AbstractFastMachine(
    itemGroup: ItemGroup,
    itemStack: SlimefunItemStack,
    recipeType: RecipeType,
    recipe: Array<out ItemStack?>,
    energyCapacity: Int,
    energyConsumption: Int,
) : MenuBlock(itemGroup, itemStack, recipeType, recipe), EnergyNetComponent {

    val recipes: List<Recipe>
        get() = _recipes
    val caches: Map<BlockPosition, FastMachineCache>
        get() = _caches

    private val _recipes = mutableListOf<Recipe>()
    private val _caches = mutableMapOf<BlockPosition, FastMachineCache>()
    private var recipeLocked = false

    private val energyCapacitySetting = IntRangeSetting(this, "energy-capacity", 0, energyCapacity, Int.MAX_VALUE)
    private val energyConsumptionSetting = IntRangeSetting(this, "energy-per-use", 0, energyConsumption, Int.MAX_VALUE)

    init {
        addItemSetting(energyCapacitySetting, energyConsumptionSetting)
    }

    fun addRecipe(recipe: Recipe) {
        check(!recipeLocked) { "Cannot add recipes after recipes are locked" }
        _recipes.add(recipe)
    }

    internal fun lockRecipes() {
        check(!recipeLocked) { "Recipes are already locked" }
        recipeLocked = true
    }

    override fun getEnergyComponentType() = EnergyNetComponentType.CONSUMER

    fun getEnergyPerUse() = if (FastMachines.configService.fmUseEnergy) energyConsumptionSetting.value else 0

    override fun getCapacity() = if (FastMachines.configService.fmUseEnergy) energyCapacitySetting.value else 0

    override fun setup(preset: BlockMenuPreset) {
        for (slot in PREVIEW_SLOTS) {
            preset.addItem(slot, ChestMenuUtils.getBackground(), ChestMenuUtils.getEmptyClickHandler())
        }
        preset.addItem(INFO_SLOT, INFO_ITEM, ChestMenuUtils.getEmptyClickHandler())
        preset.addItem(CHOICE_SLOT, NO_ITEM, ChestMenuUtils.getEmptyClickHandler())
        preset.addItem(SCROLL_UP_SLOT, SCROLL_UP_ITEM, ChestMenuUtils.getEmptyClickHandler())
        preset.addItem(SCROLL_DOWN_SLOT, SCROLL_DOWN_ITEM, ChestMenuUtils.getEmptyClickHandler())

        preset.addItem(CRAFT_SLOT, craftItem, ChestMenuUtils.getEmptyClickHandler())
        preset.addItem(
            ENERGY_SLOT,
            SlimefunHeadTexture.ENERGY_CONNECTOR.asItemStack
                .withName(" ")
                .withLore(LoreBuilder.power(getEnergyPerUse(), FastMachines.localization.getString("lores.per-craft"))),
            ChestMenuUtils.getEmptyClickHandler()
        )
    }

    override fun getInputSlots() = INPUT_SLOTS

    override fun getOutputSlots() = intArrayOf()

    abstract val craftItem: ItemStack

    abstract val recipeLoader: RecipeLoader

    override fun onNewInstance(menu: BlockMenu, b: Block) {
        val pos = BlockPosition(b)
        _caches.put(pos, FastMachineCache(this, menu))
    }

    override fun onBreak(e: BlockBreakEvent, menu: BlockMenu) {
        super.onBreak(e, menu)
        val loc = menu.location
        menu.dropItems(loc, *INPUT_SLOTS)
        _caches.remove(BlockPosition(loc))
    }

    override fun postRegister() {
        super.postRegister()
        if (!isDisabled) {
            FMRegistry.enabledFastMachines.add(this)
        }
    }

    companion object {

        internal val INPUT_SLOTS = intArrayOf(
            0, 1, 2, 3, 4, 5, 6, 7, 8,
            9, 10, 11, 12, 13, 14, 15, 16, 17,
            18, 19, 20, 21, 22, 23, 24, 25, 26,
            27, 28, 29, 30, 31, 32, 33, 34, 35,
        )
        internal val OUTPUT_SLOTS = intArrayOf(
            27, 28, 29, 30, 31, 32, 33, 34, 35,
            18, 19, 20, 21, 22, 23, 24, 25, 26,
            9, 10, 11, 12, 13, 14, 15, 16, 17,
            0, 1, 2, 3, 4, 5, 6, 7, 8,
        )
        internal val PREVIEW_SLOTS = intArrayOf(
            36, 37, 38, 39, 40, 41,
            45, 46, 47, 48, 49, 50,
        )
        internal const val SCROLL_UP_SLOT = 42
        internal const val SCROLL_DOWN_SLOT = 51
        internal const val CHOICE_SLOT = 52
        internal const val CRAFT_SLOT = 53
        internal const val INFO_SLOT = 43
        internal const val ENERGY_SLOT = 44

        internal val ITEMS_PER_PAGE = PREVIEW_SLOTS.size

        internal val NO_ITEM = FastMachines.localization.getItem("NO_ITEM", Material.BARRIER)
        internal val SCROLL_UP_ITEM = FastMachines.localization.getItem("SCROLL_UP", HeadTexture.ARROW_UP.texture)
        internal val SCROLL_DOWN_ITEM = FastMachines.localization.getItem("SCROLL_DOWN", HeadTexture.ARROW_DOWN.texture)
        internal val INFO_ITEM = FastMachines.localization.getItem("INFO", HeadTexture.INFO.texture)
    }

}
